File: clipboard-watcher.swift
// Native macOS clipboard watcher — polls NSPasteboard.changeCount
// Outputs "copy" to stdout on each clipboard change
// Much faster than osascript (~0.001ms vs ~113ms per poll)
import AppKit
import Foundation

var lastCount = NSPasteboard.general.changeCount

// Use a DispatchSource timer for precise 150ms polling
let timer = DispatchSource.makeTimerSource(queue: .main)
timer.schedule(deadline: .now(), repeating: .milliseconds(150))
timer.setEventHandler {
    let count = NSPasteboard.general.changeCount
    if count != lastCount {
        lastCount = count
        print("copy")
        fflush(stdout)
    }
}
timer.resume()

// Keep process alive
dispatchMain()

File: clipboard-watcher.js
// Standalone clipboard changeCount watcher for macOS
// Runs as a child process, sends "copy" events to parent via stdout
// Uses JXA (JavaScript for Automation) via osascript for NSPasteboard access

const { execSync } = require("child_process");

function getChangeCount() {
  try {
    return parseInt(
      execSync(
        'osascript -l JavaScript -e \'ObjC.import("AppKit"); $.NSPasteboard.generalPasteboard.changeCount\'',
        { encoding: "utf-8", timeout: 2000 }
      ).trim(),
      10
    );
  } catch {
    return -1;
  }
}

let lastCount = getChangeCount();

setInterval(() => {
  const count = getChangeCount();
  if (count !== -1 && count !== lastCount) {
    lastCount = count;
    process.stdout.write("copy\n");
  }
}, 300);
